[8.x] Track a loop variable for sequence and pass it with count to closure

jimmypuckett opened the pull request:
I have a use case where I would like to know the loop count in a sequence, so I thought that others might like the ability too.

All this does is track a loop variable similar to foreach to use with % to get the index. Then it passes it to the closure.

```php
Model::factory()->count(5)
    ->sequence(fn($loop, $count) => ['sort' => $loop * 100])
    ->create();
```

`$loop` & `$count` is optional, and people would only declare if needed.  I almost did not pass `$count`, but there are times that we do a random number in the `count` method, so there is a slight use case.

This is a re-mix of #37753 in a way to be backward compatible, as @driesvints suggested.

Thanks for considering & I hope that you see value in it.

Thanks so much for all the work that you put into Laravel!

jimmypuckett committed:
Track a loop & pass it to the closure

taylorotwell commented:
So - what is "loop"? I'm not sure you actually explain what you mean by "loop count"?

jimmypuckett commented:
@taylorotwell Sorry that the use case is not more clear.

For my specific case, I have a model `Recommendation` that `belongsTo` `Section`.  The user can drag & drop the recommendations into a specific sort order, so there is a `sort` property on `Recommendation`.  

I needed a way to increment the sort for each loop, but currently, there is not a way to know the loop count in the sequence, so we are doing.  

Here is a very simple example...

```php
$sorts = Collection::times(rand(1, 5), fn ($i) => ['sort' => ($i + 1) * 100])->toArray();

$count = count($sorts);

return $this->has(
    Section::factory()->count($count)
        ->sequence(...$sorts),
    'recommendations'
)->state(fn () => ['sort' => $sort]);
```

If sequence kept up with the loop count, then we could do...

```php
 return $this->has(
    Section::factory()->count(rand(1, 5))
        ->sequence(fn($loop) => ['sort' => ($loop +1)*100]),
    'recommendations'
);
```

I hope this is more clear.  In the end, just looking to optionally know in the sequence closure the number of times that the sequence has looped & how many times it is looping.

Thanks again for considering this.

taylorotwell commented:
Why not just make the `$count` and `$index` properties public on the Sequence class and then pass `$this` into the callback?

jimmypuckett commented:
I really like the idea of passing `$this`, but unless that I am missing something, if you only have `$index` & `$count` as they currently exist, then there is no way to know the number of times the `sequence` has run.  

Personally, I would like to repurpose `$index` as that loop count & calculate the sequence index when needed.
If you think that this adds value, we could do...

```php
<?php

namespace Illuminate\Database\Eloquent\Factories;

class Sequence
{
    /**
     * The sequence of return values.
     *
     * @var array
     */
    protected $sequence;

    /**
     * The count of the sequence items.
     *
     * @var int
     */
    public $count;

    /**
     * The current index of the sequence iteration.
     *
     * @var int
     */
    public $index = 0;

    /**
     * Create a new sequence instance.
     *
     * @param  array  $sequence
     * @return void
     */
    public function __construct(...$sequence)
    {
        $this->sequence = $sequence;
        $this->count = count($sequence);
    }

    /**
     * Get the next value in the sequence.
     *
     * @return mixed
     */
    public function __invoke()
    {
        return tap(value($this->sequence[$this->index % $this->count], $this), function () {
            $this->index = $this->index + 1;
        });
    }
}
```

Then the example above would be...

```php
return $this->has(
    Section::factory()->count(rand(1, 5))
        ->sequence(fn(Sequence $sequence) => ['sort' => ($sequence->index +1)*100]),
    'recommendations'
);
```

I would be happy to do whichever you prefer.

taylorotwell commented:
Since `$index` is protected I think it is safe to implement that suggestion. If you want to modify this PR to reflect that go ahead.

jimmypuckett commented:
Since we are going to expose a public variable, should we consider making the "API" emulate the `$loop` variable in blade with similar properties or would that be overkill?

taylorotwell commented:
Overkill IMO for now. Nobody asking for it.

jimmypuckett committed:
Make index the sequence iteration & pass sequence object back with public properties

jimmypuckett committed:
Test that index is passed to sequence closure

jimmypuckett committed:
Style

jimmypuckett commented:
OK, I have refactored the `$index` & `$count` as public & passed the `Sequence` object to the closure as we discussed.  Also, I added a simple test.

Again, thanks for considering this.
